### PR TITLE
Hw0b

### PR DESCRIPTION
**Pull request by christianjunge**
I hope I did this right!

### CODE REVIEW

File: HW0/P6/P6.txt [hunk 1]
```diff
@@ -1 +1,41 @@
...
+
+For example, if we were processing time-series data, it would be important for all of the outputs to come out in order.  
+
+
+How Much Faster?
+
+The plot P6.png shows that the ratio becomes more favorable for the Parallel process the longer the wait time is.  There is a point of diminishing returns, however, with a ratio of approximately 4, that becomes nearly flat for wait times any longer than 1/10 of a second.   The ratio of 4 for time-consuming tasks reflects that the tasks are completed 4 times as quickly by the 4 processes.  For quicker tasks, however, other considerations (communication between the processes, for instance), become more relevant and the advantage of parallel processing diminishes.  Somewhere between 10E-4 and 10E-5 seconds, the ratio crosses 1 (the point where the time for parallel and serial are equal), and for very quick tasks the serial approach is actually quicker than the parallel approach.  For these tasks in the 10E-6 to 10E-5 seconds range, the serial approach is faster because it does not incur the costs of communication between the processes and such.
```
Comment (kevinkeyjkw) [inline HW0/P6/P6.txt]: Hi Christian, great explanation on why parallel execution flattens near wait times greater than 1/10 of a second. I also learned(thanks to your explanation) that the line flattens at the 4-mark because 4 processes are executing the tasks, thus completing 4 times faster. The rest of your code is pretty well written, so not much to say about it. Good job!

**Comment by Kai-Square** [comment]
Hi Christian, thank you for your organized codes as well as clear comments! Also good explanation on the P6. Great job! Some trivial things are below.
-P4: np.diag makes it easier to create a matrix
-P5: well, my results are different from yours but i have no idea whose are desired :confused: 
-P6: i learned np.logspace from you, thanks

**Comment by evanyao27** [comment]
Hi Christian, 

Your code is very well written. The comments are really informative and the styling is wonderful. I feel like you must have some matlab experience with the slicing matrices with [:, 1] notation :) I liked it when you defined matrices with multi-line arrays -- it was very clear to read. Your written explanations are really good, except perhaps enumerating out all the possibilities is a bit too wordy. 

Great job though! 

Best, 
Evan

File: HW0/P4/P4.py [hunk 2]
```diff
@@ -40,38 +48,132 @@
...
-    # s = ?
-
+    # Propagation Matrix
+    A = np.matrix([[1, 0, 0,         dt,          0,          0],
+                   [0, 1, 0,          0,         dt,          0],
+                   [0, 0, 1,          0,          0,         dt],
+                   [0, 0, 0, (1 - c*dt),          0,          0],
+                   [0, 0, 0,          0, (1 - c*dt),          0],
+                   [0, 0, 0,          0,          0, (1 - c*dt)]])
+
```
Comment (isadoranun) [inline HW0/P4/P4.py]: It might be easier to create the A matrix in the following way:

 A = np.matrix(np.zeros([6,6]))
np.fill_diagonal(A, [1,1,1,1 - c*dt,1 - c*dt,1 - c*dt])

A[0,3] = dt
A[1,4] = dt
A[2,5] = dt

File: HW0/P4/P4.py [hunk 3]
```diff
@@ -40,38 +48,132 @@
...
+                   [ 0, by,  0,  0,  0,  0],
+                   [ 0,  0, bz,  0,  0,  0],
+                   [ 0,  0,  0,bvx,  0,  0],
+                   [ 0,  0,  0,  0,bvx,  0],
+                   [ 0,  0,  0,  0,  0,bvz]])
+
+    # Matrix C describes the dimensional stretching
+    C = np.matrix([[rx,  0,  0, 0, 0, 0],
+                   [ 0, ry,  0, 0, 0, 0],
+                   [ 0,  0, rz, 0, 0, 0]])
```
Comment (isadoranun) [inline HW0/P4/P4.py]: Same for B and C, this could save you time!
 B = np.matrix(np.zeros([6,6]))
 np.fill_diagonal(B,[bx,by,bz,bvx,bvy,bvz])

C = np.matrix(np.zeros([3,6]))
np.fill_diagonal(C,[rx,ry,rz])

File: HW0/P6/P6.txt (hunk 1 above)
Comment (isadoranun) [inline HW0/P6/P6.txt]: I agree with @kevinkeyjkw ! Very good explanation !

File: HW0/P5/P5.txt [hunk 4]
```diff
@@ -1 +1,168 @@
...
+Cashier 1 sums the bags that she is holding.  Cashier 3 is idle.  
+Second 258:
+Cashier 3 hands her bags to cashier 1.  
+Second 259:
+Cashier 3 tells cashier 1 her running total.  
+Second 260:
+Cashier 1 adds the last two numbers together.  
+
+Answer: In the first situation described, it takes 260 seconds for the cashiers to verify 256 bags.   
+
```
Comment (isadoranun) [inline HW0/P5/P5.txt]: I do not know what is the expected answer for this problem, but I agree with your reasoning.

File: HW0/P4/P4.py (hunk 3 above)
Comment (christianjunge) [inline HW0/P4/P4.py]: Wow!  Thanks, like so many things, I had no idea.

File: HW0/P5/P5.txt (hunk 4 above)
Comment (christianjunge) [inline HW0/P5/P5.txt]: Well, I got the same answers as the two people I reviewed, and they did it without writing out every step, so I guess I was on the right track.

File: HW0/P4/P4.py (hunk 2 above)
Comment (zezhouliu) [inline HW0/P4/P4.py]: This is a good suggestion is better way to create a diagonal
